### DISCUSSION TITLE
Find class with readObject methods and implements Serializable

### DISCUSSION THREAD

**Discussion by CaledoniaProject** [github/codeql#8709] [General]
I'm trying to find all readObject methods, but I'm not sure how to use extendsOrImplements():

```
predicate isDeserialize(Callable callable) {
  callable.getDeclaringType().extendsOrImplements("java.io", "Serializable") and
  (callable.hasName("readObject") or callable.hasName("readResolve"))
}
```

Can you help?

**Comment by smowton** [comment]
You should use `getASourceSupertype*()` as shown here for example: https://github.com/github/codeql/blob/main/java/ql/lib/semmle/code/java/StringFormat.qll#L72

`extendsOrImplements` checks for one-step inheritance, whereas usually you want an arbitrary ancestor or descendant type.
